### PR TITLE
More events and extensions

### PR DESCRIPTION
**Pull request by nnordrum**
Fixes #

I'll open an issue, but until I do, this is providing more metadata for functions to filter on.

## Proposed Changes
- Added event types
- Alphabetized event types
- Added properties to filter on so functions can respond to fewer events
- 🧽 Update or clean up current behavior

I am not sure if we need to do a filter on the receiver end, but this just adds them.

**Release Note**

```release-note

```

**Docs**

### CODE REVIEW

**Comment by nnordrum** [comment]
@lionelvillard  still working on #104, but figured I'd submit this separately and get all the admin out of the way.

All this does is add more metadata. Some of the fields are "duplicated", but I thought it was better to have a consistent (e.g.) `action` than have say `subject = p.Action` be "duplicated". I don't feel _really_ strongly about anything here.

The commented `Installation.ID` lines are where there are elements that are not in the API but are there.

**Comment by nnordrum** [comment]
@googlebot I signed it!

**Comment by antoineco** [comment]
Overall I think this is a good idea.

We should discuss where to strike a good balance though, because repeating a fair amount of data fields into the CE extension attributes is also not the goal of CloudEvents (this would rapidly become unmanageable and increase the payload size significantly).

@n3wscott @slinkydeveloper you're involved with CE a lot, what are your thoughts?

Maybe a longer term solution would be to allow data filtering in Knative, but it's a difficult challenge because CE data is not guaranteed to be JSON. (I digress)

File: pkg/common/events.go [hunk 1]
```diff
@@ -56,62 +58,109 @@ var ValidEvents = []gh.Event{
...
+	var exts = make(map[string]interface{})
 	var ok bool
 	switch gitHubEvent {
+	case gh.CheckRunEvent:
+		var c gh.CheckRunPayload
+		if c, ok = payload.(gh.CheckRunPayload); ok {
+			subject = strconv.FormatInt(c.CheckRun.ID, 10)
+			exts["owner"] = c.Repository.Owner.Login
+			exts["repository"] = c.Repository.Name
+			//exts["installation"] = c.Installation.ID
```
[the remark this answers is not in the document]
Comment (nnordrum) [reply pkg/common/events.go:88]: the reason they are there and commented out is the GitHub API supports them, but the client being used does not (yet). Since it's not everywhere, I added and commented out. I'm okay with removing these, but they were put in specific places for a reason (it's not present in some calls).

File: pkg/common/events.go [hunk 2]
```diff
@@ -123,149 +172,316 @@ func SubjectFromGitHubEvent(gitHubEvent gh.Event, payload interface{}, logger *z
...
 		if l, ok = payload.(gh.LabelPayload); ok {
 			// E.g., :bug: Bugfix
 			subject = l.Label.Name
+			exts["owner"] = l.Repository.Owner.Login
+			exts["repository"] = l.Repository.Name
+			//exts["installation"] = l.Installation.ID
+			exts["sender"] = l.Sender.Login
+			exts["action"] = l.Action
+
+			exts["label"] = l.Label.Name
```
[the remark this answers is not in the document]
Comment (nnordrum) [reply pkg/common/events.go:236]: was just separation "standard" things from extras. I'll remove all over.

File: pkg/common/events.go [hunk 3]
```diff
@@ -123,149 +172,316 @@ func SubjectFromGitHubEvent(gitHubEvent gh.Event, payload interface{}, logger *z
...
-		if p, ok = payload.(gh.PullRequestReviewPayload); ok {
-			subject = strconv.FormatInt(p.Review.ID, 10)
+			exts["owner"] = p.Repository.Owner.Login
+			exts["repository"] = p.Repository.Name
+			exts["installation"] = p.Installation.ID
+			exts["sender"] = p.Sender.Login
+			exts["action"] = p.Action
+
+			exts["number"] = p.Number
+			// for some reason this returns a struct not a pointer to a struct
```
[the remark this answers is not in the document]
Comment (nnordrum) [reply pkg/common/events.go:358]: pointing out an inconsistency in the API, not a mistake in coding. I can remove if you want.

File: pkg/common/events.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (antoineco) [reply pkg/common/events.go:88]: Sure but commented code is like dead code. It serves no purpose and only you knows why they are here.

**Comment by nnordrum** [comment]
I agree on the balance 100%. Where that line is, I have no idea :) I was just trying to create a intuitive way to filter things that didn't necessarily require going back to the source every time you want to add a filter (hence some of the "duplication").

The only thing I would disagree with is the use of the word "significantly" in this context. GitHub events are _very_ verbose, so adding a couple `ext` fields is not going to make a significant change in the payload size. Yes it will be larger, but you're adding _maybe_ a kb in the worse case to a message that can be up to 25MB (yes, that insanely large according to the docs). So I agree in_ principle_, just not for this particular case.

File: pkg/common/events.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (antoineco) [reply pkg/common/events.go:358]: I think yes, it's not relevant in the context, you're not doing anything particularly unsafe the reader needs to be careful about.

**Comment by antoineco** [comment]
Sorry about the confusion, when I said "significantly" I was not taking this PR specifically as an example but was rather talking about the practice of mirroring the data into the CE context in general.

No issue with this PR, as long as we agree the attributes are useful. Not sure if the "sender" is something most event consumers care about, for example, but I might be wrong since I've never done anything useful when this source myself.

**Comment by nnordrum** [comment]
I agree in theory. The issue is it _may_ be useful for somebody for like testing or something. Since they're relatively small, I went "too far" on purpose, so totally cool with removing almost anything.

**Comment by antoineco** [comment]
I'll let someone else weight in. Like I said, I'm not a user of the source myself so my capacity of judgement is pretty bad :)

File: pkg/common/handler.go [hunk 4]
```diff
@@ -95,13 +95,17 @@ func (h *Handler) handleEvent(ctx context.Context, payload interface{}, hdr http
 	h.Logger.Infof("Handling %s", gitHubEventType)
 
 	cloudEventType := sourcesv1alpha1.GitHubEventType(gitHubEventType)
-	subject := SubjectFromGitHubEvent(gh.Event(gitHubEventType), payload, h.Logger)
+	subject, extensions := SubjectAndExtensionsFromGitHubEvent(gh.Event(gitHubEventType), payload, h.Logger)
```
Comment (n3wscott) [inline pkg/common/handler.go:98]: I would have made a new method for the extensions and passed the event to it as a EventWriter or some small interface with the SetExtension method expected.

File: pkg/common/events.go [hunk 5]
```diff
@@ -56,62 +58,109 @@ var ValidEvents = []gh.Event{
...
-	gh.TeamEvent,
 	gh.TeamAddEvent,
+	gh.TeamEvent,
 	gh.WatchEvent,
 }
 
-// SubjectFromGitHubEvent computes the CE subject from GitHub event
-func SubjectFromGitHubEvent(gitHubEvent gh.Event, payload interface{}, logger *zap.SugaredLogger) string {
+// SubjectAndExtensionsFromGitHubEvent computes the CE subject from GitHub event
+func SubjectAndExtensionsFromGitHubEvent(gitHubEvent gh.Event, payload interface{}, logger *zap.SugaredLogger) (string, map[string]interface{}) {
```
Comment (n3wscott) [inline pkg/common/events.go:75]: I would revert this change, and made a new method to extract or directly set the extensions.

**Comment by n3wscott** [comment]
As to the question of when to promote payload or metadata into the cloudevent layer, it depends on what you might want to route on. The CloudEvent attributes are for routing and I could see a case where you want more info than is in the current event to route on, but I do worry about PII, like the usernames of anyone who is making github events. Those will be sent and recorded as headers. I

In the original version of this we just used what github currently has as headers and continued that.

It does increase payload size (and speed to forward the event), and yeah I agree is it not clear where you draw that line, as it might depend on the downstream usage of this. 

Do you have a usecase @nnordrum ?

File: pkg/common/events.go (hunk 5 above)
Comment (nnordrum) [reply pkg/common/events.go:75]: The reason I put them together, but I agree they arguably shouldn't be, is because different methods would use essentially the same switch statement. I can switch it to two methods.

**Comment by nnordrum** [comment]
totally agree it does increase payload size, my point was just that in this particular situation, the GitHub payloads are so large and bloated that adding 10 `ext` attributes isn't going to materially change the overall size of the payload. It _will_ make it larger, but 5.250MB and 5.251MB are basically the same as far as I'm concerned. if you really care about payload size that much, we should strip down the GitHub payloads to enough that you can find the information in GitHub, but that's also going to start another "what's actually important" conversation.

The problem for really any of the fields ultimately is, we don't know what people will find useful. As this is the first stab, I'm fine with adding/removing anything since we both are just guessing :+1:

**Comment by nnordrum** [comment]
For example, in the random PR event I'm testing with, it has `https` 236 times (switching to http would save 236 characters), the org name is used 150 times (in a forked PR; the forkee(?)'s repo is used 43 times), and the domain name and host are each used 245 times. So not saying we should just add `ext` with abandon, just in the name of "saving space", there are _MUCH_ bigger fish to fry.

**Comment by n3wscott** [comment]
> we don't know what people will find useful

The attributes on the cloudevent are for robots, not people. They are for routing decisions. So if the intent of adding this as attributes is to allow a consumer of the event to not need to parse the payload, I think we should hold off. We have an advantage over basic github webhooks where the consumer understands the shape of the payload without needing to do the standard double marshaling they might need to do before to inspect it. 

So, do you have a _routing_ usecase for this metadata? I would  rather not change the cloudevent shape without real cause or usage.

**Comment by nnordrum** [comment]
People write the robots though; that's what I meant by "people".

Let me give my specific use case that I'm working on today:
We have our repositories spread out among 3 orgs. Some of them need a linter to run when PRs are submitted.

In this use case, we would either have to: 
1) install the GitHubSource at the repo-level and lose a ton of reusability
1) Open the message every time and see if we need to lint
1) Look at a header to see the `ext[org]`, `ext[repo]`, and `ext[action]` to see if we should lint
1) write a function that basically does the predicates that aren't exposed in the `ext`

Obviously any of those would technically work, but I would rather have the metadata of org|repo|action put on the message by the `Source` and have KNative evaluate the predicate for me.

I'm totally open to other ideas/ways, this is just my understanding of how Sources/Brokers/Triggers work...

**Comment by lionelvillard** [comment]
There is a precedent: https://github.com/knative/eventing/pull/3741

**Comment by nnordrum** [comment]
> There is a precedent: [knative/eventing#3741](https://github.com/knative/eventing/pull/3741)

Interestingly, no debate on if we should there

**Comment by antoineco** [comment]
The reason for that is simple: _any_ Kubernetes object can be genetically identified by its apiVersion/kind + namespace + name, so there is not much to debate. 

GitHub events, however, don't have a similar set of attributes that are generally useful for _all_ types of event.

It's just the nature of the source that started the discussion, not the way you added it.

**Comment by nnordrum** [comment]
> GitHub events, however, don't have a similar set of attributes that are generally useful for _all_ types of event.

Agree, but bad API design doesn't mean you should just pretend the API doesn't exist. I _tried_ to be as consistent across as many objects as I could, but I concede it's not consistent, but I worked with what I had... It would be great if there was a "similar set of attributes that are generally useful for _all_ types of event", but there isn't. I'm okay paring it down to
``` go
exts["owner"] = c.Repository.Owner.Login
exts["repository"] = c.Repository.Name
//exts["installation"] = c.Installation.ID
exts["action"] = c.Action
```

as that's the closest set there is to all event types, though I would advocate for `ext["label"]` as an optional param as I feel that would be the most useful of the non-"standard" ones.

**Comment by antoineco** [comment]
Personally, I don't have anything against setting a few extensions on emitted events as long as they are useful for routing like Scott suggested (thanks for sharing your use case btw!). For instance, I can imagine it would be useful to route on `org` if you have multiple orgs and each org runs their own logic on received events (so an org can be seen as a tenant).

Regarding making the function simpler, it would be great if we could delegate everything to a sort of super-filter, but in practice all event consumers have to perform some kind of validation on what they receive, discarding a few events in production is inevitable. The syntax of filters is just not expressive enough to be a general purpose filter, it's simply there to discard event types and sources that the function obviously has no business dealing with (e.g. if your function deals with events from GitHub, for sure you want to filter events coming from, say, Zendesk, or Kafka).

If the intention is to activate the function _less_ I would suggest thinking twice. Here is why: for every trigger you create in Knative, the broker sends one copy of _each_ event it receives to it's filter. So imagine you have 100 triggers, the broker will fan out that event 100 times and each filter will perform work on a copy of the event. Not very efficient!  However, with one general purpose filter, only your function is busy, and only _once_ per event instead of 100.

**Review by n3wscott — APPROVED** [review]
I did not intend to derail the PR, i was trying to give minor feedback.

This has been sitting around for a long time, it still looks useful, we can always revert it if it causes problems. Sorry for the delay.

/lgtm
